blog/2023-06-29-automated-testing/index.mdx

```diff
@@ -0,0 +1,134 @@
+import TSCode from '!!raw-loader!/_includes/code/automated-testing.ts';
+
+As a software engineer with experience in test automation, I firmly believe in [Test-Driven Development](https://en.wikipedia.org/wiki/Test-driven_development), and more specifically, incorporating [integration testing](https://en.wikipedia.org/wiki/Integration_testing) from the very early stages of developing an application.
+
+<!-- truncate -->
+
+Integration testing helps early bug detection by verifying the interaction between different modules of a software system, which may function perfectly in isolation (as shown by unit testing), but may run into various types of issues when combined (interface/API mismatch, different data ranges etc.) Besides improving test coverage, integration testing also increases confidence when refactoring, especially when changing the implementation of a particular unit of the system under test, by proving that the system as a whole continues to function as expected.
+
+When it comes to Weaviate applications, an integration test could cover the interaction between the application code, the Weaviate server, and the inference provider (e.g. OpenAI, Cohere, Google PaLM) used via a [vectorizer](/developers/weaviate/modules/retriever-vectorizer-modules) or [generator](/developers/weaviate/modules/reader-generator-modules) module. With the advent of [embedded Weaviate](/developers/weaviate/installation/embedded), integration testing has become much easier, since Weaviate can be instantiated directly from the client code, without the need to set up and tear down a separate server.
+
+
+## Scoping the test
+
+In integration testing, it's important to delineate precisely what should be tested, in order to avoid writing tests for the _units_ of a system, and instead focus on their _integration_. In the case of a typical application built with Weaviate,
+* Search quality depends on the selected model of the [vectorization module](/developers/weaviate/modules/retriever-vectorizer-modules). This should not be the focus of integration testing. In particular, returned [distances](../distance-metrics-in-vector-search) are subject to slight changes.
+* Search itself is a core Weaviate functionality, that is tested by Weaviate's own test suite. Inserting a dummy object and searching for it would duplicate part of this test suite, and wouldn't prove anything about the integration between Weaviate and the application.
+* Integration testing should focus on automatically catching common issues including:
+  * Connection or authentication issues with the inference provider (and with Weaviate, if not using it embedded)
+  * Incomplete or incorrect data imports
+  * Using the correct number of dimensions when [bringing your own vectors](/developers/weaviate/tutorials/custom-vectors)
+  * Schema issues, like invalid class names, properties, or data types
+  * Race conditions or concurrency errors from multiple requests
+
+
+## Sample integration test
+
```

Comment:
This section seems a bit thin. It basically can be summarised as:
- we will do 1-2-3
- a big chunk of code

Can we break this down into smaller `###segments`, as you walk the reader through your code?
Maybe something like (not exactly these titles):
- Step 1 – connect to Weaviate Embedded + clean up (delete existing schemas)
- Step 2 – create the `JeopardyQuestion` collection and (add this part) assert it exists
- Step 3 – import data and assert the count

Plus we could add Step 4
- Step 4 – a query test – to verify an expected object is returned